velox/functions/sparksql/specialforms/FromJson.cpp

```diff
@@ -0,0 +1,586 @@
+        // otherwise return INCORRECT_TYPE to the caller.
+        if (isRoot) {
+          writerTyped.set_null_at(0);
+          return simdjson::SUCCESS;
+        } else {
+          return simdjson::INCORRECT_TYPE;
+        }
+      }
+      return simdjson::SUCCESS;
+    }
+  };
+
+  template <typename T>
+  static simdjson::error_code castJsonToInt(
+      Input value,
+      exec::GenericWriter& writer) {
+    SIMDJSON_ASSIGN_OR_RAISE(auto type, value.type());
+    switch (type) {
+      case simdjson::ondemand::json_type::number: {
+        SIMDJSON_ASSIGN_OR_RAISE(auto num, value.get_number());
+        switch (num.get_number_type()) {
+          case simdjson::ondemand::number_type::signed_integer:
+            return convertIfInRange<T>(num.get_int64(), writer);
+          case simdjson::ondemand::number_type::unsigned_integer:
+            return simdjson::NUMBER_OUT_OF_RANGE;
+          default:
+            return simdjson::INCORRECT_TYPE;
+        }
+      }
+      default:
+        return simdjson::INCORRECT_TYPE;
+    }
+    return simdjson::SUCCESS;
+  }
+
+  // Casts a JSON value to a float point, handling both numeric special cases
+  // for NaN and Infinity.
+  template <typename T>
+  static simdjson::error_code castJsonToFloatingPoint(
+      Input value,
+      exec::GenericWriter& writer) {
+    SIMDJSON_ASSIGN_OR_RAISE(auto type, value.type());
+    switch (type) {
+      case simdjson::ondemand::json_type::number: {
+        SIMDJSON_ASSIGN_OR_RAISE(auto num, value.get_double());
+        return convertIfInRange<T>(num, writer);
+      }
+      case simdjson::ondemand::json_type::string: {
+        SIMDJSON_ASSIGN_OR_RAISE(auto s, value.get_string());
+        constexpr T kNaN = std::numeric_limits<T>::quiet_NaN();
+        constexpr T kInf = std::numeric_limits<T>::infinity();
+        if (s == "NaN") {
+          writer.castTo<T>() = kNaN;
+        } else if (s == "+INF" || s == "+Infinity" || s == "Infinity") {
+          writer.castTo<T>() = kInf;
+        } else if (s == "-INF" || s == "-Infinity") {
+          writer.castTo<T>() = -kInf;
+        } else {
+          return simdjson::INCORRECT_TYPE;
+        }
+        break;
+      }
+      default:
+        return simdjson::INCORRECT_TYPE;
+    }
+    return simdjson::SUCCESS;
+  }
+
+  template <typename To, typename From>
+  static simdjson::error_code convertIfInRange(
+      From x,
+      exec::GenericWriter& writer) {
+    static_assert(std::is_signed_v<From> && std::is_signed_v<To>);
+    if constexpr (sizeof(To) < sizeof(From)) {
+      constexpr From kMin = std::numeric_limits<To>::lowest();
+      constexpr From kMax = std::numeric_limits<To>::max();
+      if (!(kMin <= x && x <= kMax)) {
+        return simdjson::NUMBER_OUT_OF_RANGE;
+      }
+    }
+    writer.castTo<To>() = x;
+    return simdjson::SUCCESS;
+  }
+
+  static folly::F14FastMap<std::string, int32_t> makeFieldIndicesMap(
+      const RowType& rowType,
+      bool allFieldsAreAscii) {
```

Comment:
nit: perhaps rename as 'asLower' and document this function.

Reply:
Added the comments, thanks.